cpp/src/groupby/sort/group_dense_rank_scan.cu

```diff
@@ -0,0 +1,112 @@
+    flat_order = table_view{{order_by}};
+  }
+  auto const d_flat_order = table_device_view::create(flat_order, stream);
+  row_equality_comparator<has_nested_nulls> comparator(*d_flat_order, *d_flat_order, true);
+  auto ranks = make_fixed_width_column(
+    data_type{type_to_id<size_type>()}, order_by.size(), mask_state::ALL_VALID, stream, mr);
+  auto mutable_ranks = ranks->mutable_view();
+
+  if (order_by.type().id() == type_id::STRUCT && order_by.has_nulls()) {
+    auto const d_col_order = column_device_view::create(order_by, stream);
+    thrust::tabulate(rmm::exec_policy(stream),
+                     mutable_ranks.begin<size_type>(),
+                     mutable_ranks.end<size_type>(),
+                     [comparator,
+                      d_col_order = *d_col_order,
+                      labels      = group_labels.data(),
+                      offsets     = group_offsets.data()] __device__(size_type row_index) {
+                       if (row_index == offsets[labels[row_index]]) { return true; }
+                       bool const lhs_is_null{d_col_order.is_null(row_index)};
+                       bool const rhs_is_null{d_col_order.is_null(row_index - 1)};
+                       if (lhs_is_null && rhs_is_null) {
+                         return false;
+                       } else if (lhs_is_null != rhs_is_null) {
+                         return true;
+                       }
+                       return !comparator(row_index, row_index - 1);
+                     });
+
+  } else {
+    thrust::tabulate(
+      rmm::exec_policy(stream),
+      mutable_ranks.begin<size_type>(),
+      mutable_ranks.end<size_type>(),
+      [comparator, labels = group_labels.data(), offsets = group_offsets.data()] __device__(
+        size_type row_index) {
+        return row_index == offsets[labels[row_index]] || !comparator(row_index, row_index - 1);
+      });
+  }
+
+  thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                group_labels.begin(),
+                                group_labels.end(),
+                                mutable_ranks.begin<size_type>(),
+                                mutable_ranks.begin<size_type>());
+  return ranks;
+}
+}  // namespace
+std::unique_ptr<column> dense_rank_scan(column_view const& order_by,
+                                        device_span<size_type const> group_labels,
+                                        device_span<size_type const> group_offsets,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+{
+  if (order_by.type().id() == type_id::STRUCT) {
+    bool nested_nulls =
+      std::any_of(order_by.child_begin(), order_by.child_end(), [](auto const& col) {
+        return has_nested_nulls(table_view{{col}});
+      });
+    if (nested_nulls) {
+      return generate_dense_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
+    }
```

Comment:
I don't know why you don't check nulls for the `order_by` column but only check nulls for its children?
```suggestion
  if (order_by.type().id() == type_id::STRUCT && has_nested_nulls(table_view{std::vector<column_view>{order_by.child_begin(), order_by.child_end()}}) {
      return generate_dense_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
```

If `order_by` does not have nulls, just use this:
```
  if (order_by.type().id() == type_id::STRUCT && has_nested_nulls(table_view{{order_by}}) {
      return generate_dense_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
```

Maybe we can combine with the line below to just this:
```
  if ((order_by.type().id() == type_id::STRUCT && has_nested_nulls(table_view{{order_by}})) 
      || order_by.has_nulls()) {
      return generate_dense_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
} 
return generate_dense_ranks<false>(order_by, group_labels, group_offsets, stream, mr);
```